pkg/rpc/server/server.go

```diff
@@ -1688,12 +1726,23 @@ func (s *Server) getFakeNextBlock() (*block.Block, error) {
 // witness invocation script in case of `verification` trigger (it pushes `verify`
 // arguments on stack before verification). In case of contract verification
 // contractScriptHash should be specified.
-func (s *Server) runScriptInVM(t trigger.Type, script []byte, contractScriptHash util.Uint160, tx *transaction.Transaction, verbose bool) (*result.Invoke, *response.Error) {
-	b, err := s.getFakeNextBlock()
-	if err != nil {
-		return nil, response.NewInternalServerError("can't create fake block", err)
+func (s *Server) runScriptInVM(t trigger.Type, script []byte, contractScriptHash util.Uint160, tx *transaction.Transaction, b *block.Block, verbose bool) (*result.Invoke, *response.Error) {
+	var (
+		err error
+		ic  *interop.Context
+	)
+	if b == nil {
+		b, err = s.getFakeNextBlock()
+		if err != nil {
+			return nil, response.NewInternalServerError("can't create fake block", err)
+		}
+		ic = s.chain.GetTestVM(t, tx, b)
+	} else {
+		ic, err = s.chain.GetTestHistoricVM(t, tx, b)
```

Comment:
Passing an existing block is not correct. We just want state at height X, but the block in the context should be a fake one as in regular test invocation.